lib/backend/report.go

```diff
@@ -322,6 +325,49 @@ func (s *Reporter) ConditionalDelete(ctx context.Context, key []byte, revision s
 	return err
 }
 
+func (s *Reporter) AtomicWrite(ctx context.Context, condacts []ConditionalAction) (revision string, err error) {
+	// note: the atomic write method's metrics are counted toward both the general 'write'
+	// metrics as well as equivalent metrics specific to atomic write.
+	ctx, span := s.Tracer.Start(
+		ctx,
+		"backend/AtomicWrite",
+		oteltrace.WithAttributes(
+			attribute.Int("condacts", len(condacts)),
+		),
+	)
+	defer span.End()
+
+	start := s.Clock().Now()
+	revision, err = s.Backend.AtomicWrite(ctx, condacts)
+
+	elapsed := s.Clock().Since(start).Seconds()
+	writeLatencies.WithLabelValues(s.Component).Observe(elapsed)
+	atomicWriteLatencies.WithLabelValues(s.Component).Observe(elapsed)
+
+	writeRequests.WithLabelValues(s.Component).Inc()
+	atomicWriteRequests.WithLabelValues(s.Component).Inc()
+	if err != nil {
+		if errors.Is(err, ErrConditionFailed) {
+			atomicWriteConditionFailed.WithLabelValues(s.Component).Inc()
+		} else {
+			writeRequestsFailed.WithLabelValues(s.Component).Inc()
+			atomicWriteRequestsFailed.WithLabelValues(s.Component).Inc()
+		}
```

Comment:
I've gone back and forth on this.  On one hand, I think that `ConditionFailed` is the atomic write API working as intended, and counting them toward failure metrics might be misleading.  Especially since you generally expect the value of request failure metrics to be zero unless something is truly wrong (poor connectivity, throttling, etc).  On the other hand, a large number of `ConditionFailed` errors in a short amount of time probably _is_ symptomatic of a bug, and counting `ConditionFailed` toward the standard request failure metrics might increase visibility.  I currently lean toward leaving as-is, but not strongly.

Looking at our other "optimistic locking" methods, we are not consistent about how they treat failed preconditions.  `CompareAndSwap` _doesn't_ count a `CompareFailed` as a failed request, but `ConditionalUpdate` does treat `ErrIncorrectRevision` as a failure.  So historically, optimistic locking failure hasn't resulted in failure metrics, but as of the introduction of revisions it is now possible.

Given the above, do you have strong feelings either way?  I think probably the most important part will actually be to update one of the two existing methods so that all optimistic locking is treated the same, regardless of which way we go.